### ISSUE TITLE
`IB.client.getReqId()` returns already used ID

### ISSUE THREAD

**Issue by radekwlsk** [erdewit/ib_insync#80]
I am using `IB.client.getReqId()` to place my version of bracket orders and set `parentId` in target and stop loss orders.

But I have stumbled upon an error:

```
Traceback (most recent call last):
  (...)
  File "_", line 53, in bracket_order
    parent_trade: Trade = ib_client.placeOrder(contract, parent)
  File "/venv/lib/python3.6/site-packages/ib_insync/ib.py", line 528, in placeOrder
    assert trade.orderStatus.status not in OrderStatus.DoneStates
AssertionError
```

Which means that trade with such ID was found.

Parent order is created like that:
```
parent = MarketOrder(
    action, quantity,
    orderId=ib_client.client.getReqId(),
    transmit=False,
    conditionsIgnoreRth=False,
    outsideRth=False
)
```

Entire `bracket_order` function is copy of `ib_insync.ib.IB.bracketOrder` source with parent placed as `MarketOrder` instead of `LimitOrder`. So all but `stopLoss` are `transmit=False`. Orders are placed in order: 
  - parent, 
  - target, 
  - stop.

Is there a problem with `IB` that it start `IB.client` on wrong ID or does not update it properly?

**Comment by erdewit** [comment]
>     assert trade.orderStatus.status not in OrderStatus.DoneStates
> AssertionError
> Which means that trade with such ID was found.

No, it means that the order that is submitted is already "done" (filled, canceled or inactive). A duplicate orderId will give an IB error message and not an exception.

**Comment by radekwlsk** [comment]
@erdewit can you tell my why the issue is closed? I am jusy creating the `MarketOrder`, using clients method get id that should be unique and then call `IB.placeOrder` with it. How can it be filled, cancelled or inactive when it was never at TWS before? There is an issue with the library and it is not resolved. Reopen the issue please.

**Comment by erdewit** [comment]
Can you supply the minimum code to replicate the issue.

**Comment by radekwlsk** [comment]
@erdewit I can on Monday, but really all I did was copy `IB.bracketOrder` and made parent a `MarketOrder`, placing them in the function before returning in order mentioned before (same as `IB.bracketOrder`).

**Comment by erdewit** [comment]
From this I cannot see anything. But as I already mentioned the assertion error that you get is not related to the reqId.

**Comment by radekwlsk** [comment]
@erdewit Note that
```
assert trade.orderStatus.status not in OrderStatus.DoneStates
```
in error trace indicates that in 
`placeOrder(self, contract: Contract, order: Order) -> Trade`
from
```
        orderId = order.orderId or self.client.getReqId()
        self.client.placeOrder(orderId, contract, order)
        now = datetime.datetime.now(datetime.timezone.utc)
        key = self.wrapper.orderKey(
                self.wrapper.clientId, orderId, order.permId)
        trade = self.wrapper.trades.get(key)
```
`trade` was received and `if trade:` was matched: `# this is a modification of an existing order`

**Comment by erdewit** [comment]
Okay, I see now. But I still can't do anything about it without logs that show what's happening or some code to reproduce it.

**Comment by erdewit** [comment]
>  Is there a possibility that Client.getReqId() called multiple times in a short notice fails to update the sequence? As it is async may be the case that two calls return the same id?

If you're not multithreading then no, it's not possible. The only way I can see the problem occurring is if you use threading or directly use client._reqIdSeq somewhere.

**Comment by radekwlsk** [comment]
@erdewit I am using (APScheduler)[https://apscheduler.readthedocs.io/en/latest/index.html] with asyncio scheduler (`apscheduler.schedulers.asyncio.AsyncIOScheduler`) but there is single trades placement job and places trades in a sequence one by one. The `IB` client instance for trades placement is global for the scheduler, shared between the jobs. May there by any issue with that approach?

**Comment by erdewit** [comment]
I believe the AsyncIOScheduler should not pose a problem. But you'll have to debug this situation yourself, I would suggest to log everything that happens and see what orders clash with each other.

**Comment by 3rock618** [comment]
Had this same issue. Mine occurred right after an IBGW error where I had to restart IBGW and reconnect all the clients. All my trade data was still stored in `ib.trades`, but perhaps the reconnection effects the `ib.client._reqIdSeq` somehow. 

In any case, I'm going to patch client.getReqId:

```python
def getReqId(self) -> int:
    """Get new request ID."""
    if not self.isReady():
        raise ConnectionError('Not connected')
    retry = bool('first attempt')
    while retry:
        retry = self._reqIdSeq in [k[1] for k in self.wrapper.trades if isinstance(k,tuple)]
        newId = self._reqIdSeq
        self._reqIdSeq += 1
    return newId
```

**Comment by erdewit** [comment]
In the upcoming version of ib_insync the maximum orderId of open/completed orders is used to set the reqId sequence (see [this code](https://github.com/erdewit/ib_insync/blob/83e6e9aff4716bd3bb1b30816fb3ad21e23ec8fd/ib_insync/ib.py#L1647)).

It seems the "nextValidId" from IB can't be relied upon when TWS/gateway gets restarted.

**Comment by rajeshon21** [comment]
>the maximum orderId of open/completed orders is used to set the reqId sequence
It looks like I ran into a variation of Canceled/Rejected orders which now seems to be causing duplicate orders error. I was developing some new code, and exercising it with TWS, but code was mistakenly not providing account number, so those orders got rejected by TWS. Below are log entries for one such failed order. Then when I fixed the bug and ran the code again, it basically started giving Duplicate order error

Initial Failed logs because of bug -

BracketOrder(parent=StopOrder(orderId=188, action='BUY', totalQuantity=255.0, auxPrice=126.34, tif='GTC', transmit=False), takeProfit=LimitOrder(orderId=189, action='SELL', totalQuantity=255.0, lmtPrice=131.0, tif='GTC', transmit=False, parentId=188, outsideRth=True), stopLoss=StopOrder(orderId=190, action='SELL', totalQuantity=255.0, auxPrice=118.5, tif='GTC', parentId=188))
Error 201, reqId 188: Order rejected - reason:Invalid or missing IBCustAcctNo
Canceled order: Trade(contract=Stock(symbol='PG', exchange='SMART', currency='USD'), order=StopOrder(orderId=188, clientId=2, permId=241894548, action='BUY', totalQuantity=255.0, lmtPrice=126.37, auxPrice=126.34, tif='GTC', transmit=False, usePriceMgmtAlgo=True), orderStatus=OrderStatus(orderId=188, status='Cancelled', filled=0.0, remaining=255.0, avgFillPrice=0.0, permId=241894548, parentId=0, lastFillPrice=0.0, clientId=2, whyHeld='', mktCapPrice=0.0), fills=[], log=[TradeLogEntry(time=datetime.datetime(2021, 3, 7, 18, 27, 23, 995821, tzinfo=datetime.timezone.utc), status='PendingSubmit', message=''), TradeLogEntry(time=datetime.datetime(2021, 3, 7, 18, 27, 26, 180034, tzinfo=datetime.timezone.utc), status='Inactive', message=''), TradeLogEntry(time=datetime.datetime(2021, 3, 7, 18, 27, 26, 275944, tzinfo=datetime.timezone.utc), status='Cancelled', message='Error 201, reqId 188: Order rejected - reason:Invalid or missing IBCustAcctNo')])


After fixing bu to include account number, it gives Duplicate Order error [I changed account field to XXX]

BracketOrder(parent=StopOrder(orderId=188, action='BUY', totalQuantity=255.0, auxPrice=126.34, tif='GTC', transmit=False), takeProfit=LimitOrder(orderId=189, action='SELL', totalQuantity=255.0, lmtPrice=131.0, tif='GTC', transmit=False, parentId=188, outsideRth=True), stopLoss=StopOrder(orderId=190, action='SELL', totalQuantity=255.0, auxPrice=118.5, tif='GTC', parentId=188))
Error 103, reqId 188: Duplicate order id
Canceled order: Trade(contract=Stock(symbol='PG', exchange='SMART', currency='USD'), order=StopOrder(orderId=188, clientId=2, action='BUY', totalQuantity=255.0, auxPrice=126.34, tif='GTC', transmit=False, account='XXX', usePriceMgmtAlgo=True), orderStatus=OrderStatus(orderId=188, status='Cancelled', filled=0, remaining=0, avgFillPrice=0.0, permId=0, parentId=0, lastFillPrice=0.0, clientId=0, whyHeld='', mktCapPrice=0.0), fills=[], log=[TradeLogEntry(time=datetime.datetime(2021, 3, 7, 18, 29, 2, 457230, tzinfo=datetime.timezone.utc), status='PendingSubmit', message=''), TradeLogEntry(time=datetime.datetime(2021, 3, 7, 18, 29, 2, 459230, tzinfo=datetime.timezone.utc), status='Cancelled', message='Error 103, reqId 188: Duplicate order id')])

I restarted TWS after this and tried again, but still the same. So not quite sure how to get past this now.

**Comment by nemozny** [comment]
> It seems the "nextValidId" from IB can't be relied upon when TWS/gateway gets restarted.

I ran into these problems with a different API client too and it really is an issue of IB.
The only way to avoid getting "duplicate orderIds" is to track orderIds on your own and use the highest number you'll get:
```
orderIdFromIB = IB.client.getReqId()
orderIdSaved = DB.getOrderId()
orderId = orderIdFromIB if orderIdFromIB > orderIdSaved else orderIdSaved

def getOrderId:
   return orderId += 1
```
I my experience IB was lagging behind my number by several hundreds (for example 15800 was the last orderId I used/tracked vs 15000 I got from IB).
I think "service requests" are to blame, like requesting historical data, requesting executions, etc. I think IB is failing to track these. They report only orderIds used in previously placed orders and that creates these gaps.

What also works is to reset the orderId sequence, but that will solve only the immediate problem ... not a system solution.
![image](https://user-images.githubusercontent.com/1666780/114055139-c6071e80-9890-11eb-9505-80287d6d263c.png)

**Comment by rajeshon21** [comment]
@nemozny Thanks for the info. In my case, I have found it mainly happening after previously placed orders have been canceled. OK, I will look to save the orderId and use it

**Comment by mattsta** [comment]
Note: the IBKR API is actually fine here, but there's a race condition in the ib_insync startup where the wrong ID is picked most of the time (if you have canceled orders or market data requests on startup).

The fix in https://github.com/erdewit/ib_insync/issues/308#issuecomment-817395866 is to wait until the proper "next order id" API request (command 9) is received instead of completing startup using only the highest currently visible order id.

**Comment by mattsta** [comment]
Oh, on second look, this was actually recently fixed in https://github.com/erdewit/ib_insync/commit/eac5b1de40bcc4a58fc646c353d3b289ccf872fd but I just didn't notice the new release had it fixed.

The 0.9.66 release from April 26 has fixed the duplicate order ID problem now.

**Comment by nemozny** [comment]
Thanks @mattsta, but if it was fixed I wouldn't keep seeing the error. I started tracking orderId independently from what IB or ib_insync returns and I'll try and report back after some time.

**Comment by mattsta** [comment]
Oh, weird.

You can get the exact correct ID from the upstream API using `self.ib.client.reqIds(0)` (where `self.ib` is your connection) but note the return value will be in the console of the gateway (under `Client 0` for example) because for some reason ib_insync isn't capturing the return value properly.

So run `.reqIds(0)` then look at the very bottom of the gateway `Client 0` tab for something like:

```haskell
[date] -> ---           9-1-XXXX-
```

where `XXXX` will be your current next good order ID.  Then compare the in-gateway value with the live value of `self.ib.client._reqIdSeq`.

I spent a few hours tracking down the duplicate ID problem a while ago, and I'm fairly certain the 0.9.66 release fixes it on startup, but double checking with the actual API call will tell you if it's good or not.

**Comment by rajeshon21** [comment]
I have updated to 0.9.66 after I had put a fix of my own to keep track of the order ids locally, I am fairly certain the issue is now resolved, but I will check again after removing my own fixes.

**Comment by nemozny** [comment]
I can confirm the issue seems to be fixed with 0.9.66. I started tracking for orderId collisions, but since the update I haven't seen any.
In my previous comment I was NOT on 0.9.66.
